### PR TITLE
Add storage unit tests

### LINKED ISSUE #263: Harness should support storage-get and storage-add
It doesn't currently expose ways of interacting with storage.

People wanting to test their charms are going to want that interaction.

### PR DESCRIPTION
**Pull request by phvalguima**
* Implements storage_add, storage_list, storage_get in testing.py backend;
* Implements add_storage to be used to write unit tests.

---

As specified in Juju documentation: https://juju.is/docs/sdk/storage
StorageAttachedEvents should be called prior to InstallEvent. That is accomplished by emitting a storage_attached events just before install on begin_with_initial_hooks:

```
        for storage_name in self._meta.storages:
            if len(self._backend.storage_list(storage_name)) > 0:
                storage_name = storage_name.replace('-', '_')
                self._charm.on[storage_name].storage_attached.emit()
        # Storage done, emit install event
        self._charm.on.install.emit()
```

---

Examples:
The user can run add_storage to attach a number of disks to the charm unit.

```
...
        harness.add_storage("data", count=2)
        harness.update_config({
             ...
        })
        self.harness.begin_with_initial_hooks()
```

Fixes #263

### CODE REVIEW

File: ops/testing.py [hunk 1]
```diff
@@ -874,6 +898,16 @@ def __init__(self, unit_name, meta):
         self._unit_status = {'status': 'maintenance', 'message': ''}
         self._workload_version = None
         self._resource_dir = None
+        # Format:
+        # { "storage_name": {"<ID1>": { <other-properties> }, ... }
+        # <ID1>: device id that is key for given storage_name
+        # Initialize the _storage_list with values present on metadata.yaml
+        self._storage_list = \
+            {k: {} for k, v in self._meta.storages.items()}
```
Comment (rbarry82) [inline ops/testing.py:906]: Nit -- our linters are ok with your line being longer. No need to break it up

File: ops/testing.py [hunk 2]
```diff
@@ -977,13 +1011,22 @@ def status_set(self, status, message='', *, is_app=False):
             self._unit_status = {'status': status, 'message': message}
 
     def storage_list(self, name):
-        raise NotImplementedError(self.storage_list)
+        return [id for id, props in self._storage_list[name].items()]
```
Comment (rbarry82) [inline ops/testing.py:1014]: Why not `[id for id in self._storage_list[name].keys()]` ?

**Review by rbarry82 — APPROVED** [review]
Two small comments, but otherwise LGTM.

Are we adding tests for these tests?

File: ops/testing.py (hunk 2 above)
Comment (benhoyt) [reply ops/testing.py:1014]: Or `list(self._storage_list[name].keys())`?

**Review by benhoyt — APPROVED** [review]
Couple of style comments, but LGTM

**Review by jameinel — COMMENTED** [review]
This only handles adding storage before init, and only begin_with_initial_hooks(). Ideally it would support a bit more of the storage model (being able to remove storage, etc).
But I do think it covers the biggest use cases for storage so is still worth landing.
I would like to at least see a "if self.charm: trigger storage hook" in the add_storage function.

**Comment by jnsgruk** [comment]
@rbarry82, let's land this PR (with suggestions made by yourself and @benhoyt), but please add the points from John about storage events later in the charm's lifecycle to the Operator Framework backlog.

**Review by jnsgruk — APPROVED** [review]

**Comment by rbarry82** [comment]
@phvalguima update to match the comments?

**Pull request merged**